.github/workflows/ci.yml

```diff
@@ -0,0 +1,56 @@
+name: CI
+
+on: [push, pull_request, release]
+
+jobs:
+  qa:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: pre-commit/action@v3.0.0
+
+  check-links:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: gaurav-nelson/github-action-markdown-link-check@v1
+        name: Check links in markdown files
+        with:
+          use-quiet-mode: "yes"
+          use-verbose-mode: "yes"
+
+  build:
+    name: Build all projects
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [ubuntu-latest, windows-latest, macos-latest]
+
+    steps:
+      - uses: actions/checkout@v3
+
+      - name: Install dependencies (Ubuntu)
+        if: matrix.os == 'ubuntu-latest'
+        run: |
+          sudo apt update
+          sudo apt upgrade
+          sudo apt install build-essential cmake ninja-build
+
+      - name: Set up MSVC (Windows)
+        if: matrix.os == 'windows-latest'
+        uses: ilammy/msvc-dev-cmd@v1
+        with:
+          arch: x64
+
+      - name: Install dependencies (macOS)
+        if: matrix.os == 'macos-latest'
+        run: brew install cmake ninja
```

Comment:
I think `cmake` is included by default in macos runners: https://github.com/actions/runner-images/blob/main/images/macos/macos-13-Readme.md

Reply:
What's `Ninja`, by the way? Would the potential target audience of the course know about it?

Reply:
I deliberately omitted mention of it from the README, because I don't want students to try to use it. They should be able to build things just by following the instructions.

It's basically a cross-platform version of `make` that's fast and has a less verbose output. It's useful for the CI system because it runs in parallel by default and also works with Windows. In the past I've found that CMake generates Visual Studio project files on Windows by default, but maybe that's no longer true (or maybe it would work from the command line anyway).

https://ninja-build.org/